File: app/src/main/java/com/quranapp/android/utils/tafsir/TafsirUtils.java
package com.quranapp.android.utils.tafsir;

import androidx.annotation.Nullable;

import com.quranapp.android.api.models.tafsir.TafsirInfoModel;
import com.quranapp.android.utils.app.AppUtils;
import com.quranapp.android.utils.reader.tafsir.TafsirManager;
import com.quranapp.android.utils.univ.FileUtils;

import java.util.List;
import java.util.Map;

public class TafsirUtils {
    public static final String DIR_NAME = FileUtils.createPath(AppUtils.BASE_APP_DOWNLOADED_SAVED_DATA_DIR, "tafsirs");
    public static final String AVAILABLE_TAFSIRS_FILENAME = "available_tafsirs.json";
    public static final String KEY_TAFSIR = "key.tafsir";

    @Nullable
    public static String getTafsirName(String key) {
        if (key == null) {
            return null;
        }

        TafsirInfoModel model = TafsirManager.getModel(key);
        if (model == null) {
            return null;
        }

        return model.getName();
    }

    public static String getTafsirSlugFromKey(String key) {
        TafsirInfoModel model = TafsirManager.getModel(key);
        if (model == null) {
            return null;
        }

        return model.getSlug();
    }

    public static String getDefaultTafsirKey() {
        Map<String, List<TafsirInfoModel>> models = TafsirManager.getModels();
        if (models == null) {
            return null;
        }

        List<TafsirInfoModel> tafsirs = models.get("en");

        if (tafsirs == null || tafsirs.isEmpty()) {
            return null;
        }

        return tafsirs.get(0).getKey();
    }

    public static boolean isUrdu(String key) {
        TafsirInfoModel model = TafsirManager.getModel(key);
        if (model == null) {
            return false;
        }

        return model.getLangCode().equals("ur");
    }
}

File: app/src/main/java/com/quranapp/android/utils/tafsir/TafsirUtils.kt
package com.quranapp.android.utils.tafsir

import com.quranapp.android.api.models.tafsir.TafsirModel
import com.quranapp.android.compose.utils.appFallbackLanguageCodes
import com.quranapp.android.utils.app.AppUtils
import com.quranapp.android.utils.reader.tafsir.TafsirManager
import com.quranapp.android.utils.univ.FileUtils

object TafsirUtils {
    @JvmField
    val DIR_NAME: String =
        FileUtils.createPath(AppUtils.BASE_APP_DOWNLOADED_SAVED_DATA_DIR, "tafsirs")

    const val AVAILABLE_TAFSIRS_FILENAME = "available_tafsirs_v2.json"
    const val KEY_TAFSIR = "key.tafsir"

    fun getDefaultTafsirKey(): String? {
        val models = TafsirManager.getModels()?.takeIf { it.isNotEmpty() } ?: return null

        val idealModels = appFallbackLanguageCodes().firstNotNullOfOrNull {
            models[it]
        } ?: models["en"]

        return idealModels?.firstOrNull()?.key
    }

    fun isUrdu(key: String): Boolean {
        val model = TafsirManager.getModel(key) ?: return false
        return model.langCode == "ur"
    }

    fun isArabic(key: String): Boolean {
        val model = TafsirManager.getModel(key) ?: return false
        return model.langCode == "ar"
    }

    fun tafsirVerseRangeInChapter(tafsir: TafsirModel, chapterNo: Int): IntRange? {
        val nums = tafsir.verses.mapNotNull { key ->
            val parts = key.split(":")
            if (parts.size < 2) return@mapNotNull null
            val surah = parts[0].toIntOrNull() ?: return@mapNotNull null
            val ayah = parts[1].toIntOrNull() ?: return@mapNotNull null
            if (surah == chapterNo) ayah else null
        }
        if (nums.isEmpty()) return null
        val sorted = nums.sorted()
        return sorted.first()..sorted.last()
    }
}
